pkg/tecdsa/dkg/member.go

```diff
@@ -256,9 +259,382 @@ func (fm *finalizingMember) MarkInactiveMembers(
+			)
+			continue
+		}
+
+		receivedValidResultSignatures[message.senderID] = message.signature
+	}
+
+	// Register member's self signature.
+	receivedValidResultSignatures[sm.index] = sm.selfDKGResultSignature
+
+	return receivedValidResultSignatures, nil
+}
+
+// shouldAcceptMessage indicates whether the given member should accept
+// a message from the given sender.
+func (sm *signingMember) shouldAcceptMessage(
+	senderID group.MemberIndex,
+	senderPublicKey []byte,
+) bool {
+	isMessageFromSelf := senderID == sm.index
+	isSenderValid := sm.membershipValidator.IsValidMembership(
+		senderID,
+		senderPublicKey,
+	)
+	isSenderAccepted := sm.group.IsOperating(senderID)
+
+	return !isMessageFromSelf && isSenderValid && isSenderAccepted
+}
+
+// SubmissionConfig contains parameters describing DKG submission process.
+type SubmissionConfig struct {
+	// GroupSize is the size of a group in TBTC.
+	GroupSize int
+	// HonestThreshold is the minimum number of active participants behaving
+	// according to the protocol needed to generate a signature.
+	HonestThreshold int
+	// ResultPublicationBlockStep is the duration (in blocks) that has to pass
+	// before group member with the given index is eligible to submit the
+	// result.
+	// Nth player becomes eligible to submit the result after
+	// T_dkg + (N-1) * T_step
+	// where T_dkg is time for phases 1-12 to complete and T_step is the result
+	// publication block step.
+	ResultPublicationBlockStep uint64
+}
+
+// SubmittingMember represents a member submitting a DKG result to the
+// blockchain along with signatures received from other group members supporting
+// the result.
+type submittingMember struct {
+	logger log.StandardLogger
+
+	// Represents the member's position for submission.
+	index  group.MemberIndex
+	config *SubmissionConfig
+}
+
+// SubmitDKGResult sends a result, which contains the group public key and
+// signatures, to the chain.
+// It checks if the result has already been published to the blockchain by
+// checking if a group with the given public key is already registered. If not,
+// it determines if the current member is eligible to submit a result.
+// If allowed, it submits the result to the chain.
+// A user's turn to publish is determined based on the user's index and block
+// step.
+// If a result is submitted by another member and it's accepted by the chain,
+// the current member finishes the phase immediately, without submitting
+// their own result.
+// It returns the on-chain block height of the moment when the result was
+// successfully submitted on chain by the member. In case of failure or result
+// already submitted by another member it returns `0`.
+func (sm *submittingMember) SubmitDKGResult(
+	result *Result,
+	signatures map[group.MemberIndex][]byte,
+	resultSubmitter ResultSubmitter,
+	blockCounter chain.BlockCounter,
+	startBlockHeight uint64,
+) error {
+	// TODO: Consider moving the whole implementation to the resultSubmitter
+	//       interface
```

Comment:
That's right, the logic of submission should be controlled by the application performing DKG (so `pkg/tbtc` in our case).

Reply:
Done in e3e4bcd23c969076d0a7a55673f4c1dbc8dd38a3.